src/sqlfluff/core/plugin/hookspecs.py

```diff
@@ -14,15 +16,24 @@ class PluginSpec:
 
     @hookspec
     @abstractmethod
-    def get_rules(self):
-        """Get plugin rules."""
+    def get_rules(self) -> List[Type[RuleType]]:
+        """Get plugin rules.
+
+        NOTE: While the type annotation for this method returns `Type[RuleType]`
+        all plugin implementations should instead import `sqlfluff.core.rules` and
+        return the more specific `Type[BaseRule]`. This base definition of the
+        PluginSpec cannot import that yet as it would cause a circular import.
+        """
 
     @hookspec
     @abstractmethod
-    def load_default_config(self) -> dict:
+    def load_default_config(self) -> Dict[str, Any]:
         """Loads the default configuration for the plugin."""
 
     @hookspec
     @abstractmethod
-    def get_configs_info(self) -> dict:
+    # TODO: This type annotation could probably be more specific but that would
+    # require making the config info object something more like a namedTuple rather
+    # than a dict.
+    def get_configs_info(self) -> Dict[str, Dict[str, Any]]:
```

Comment:
I'm not proposing that I resolve this TODO in this PR. It's more of a marker for later improvements. What I've done here is already better than what was here before.